### PR TITLE
:seedling: [WIP] Runtimesdk/validate handlers on discovery

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Add first round of validation rules on discovery of extensions in the extension controller.

### CODE REVIEW

File: exp/runtime/internal/controllers/extension_controller.go [hunk 1]
```diff
@@ -197,14 +200,59 @@ func (r Reconciler) discoverExtension(ctx context.Context, extension *runtimev1.
 // if any of these checks fails the response is invalid and an error is returned. Extensions with previously valid
 // RuntimeExtension registrations are not removed from the registry or the object's status.
 func validateRuntimeExtensionDiscovery(ext *runtimev1.Extension) error {
+	names := make(map[string]bool)
+	var errs []error
+	// TODO: Implement the same validation rules in the runtime SDK server side Discovery handler.
 	for _, runtimeExtension := range ext.Status.RuntimeExtensions {
```
Comment (ykakarap) [inline exp/runtime/internal/controllers/extension_controller.go:206]: How about also adding a validation rule to make sure that the hook is valid?

File: exp/runtime/internal/controllers/extension_controller.go [hunk 2]
```diff
@@ -197,14 +200,59 @@ func (r Reconciler) discoverExtension(ctx context.Context, extension *runtimev1.
...
+		}
+
+		names[runtimeExtension.Name] = true
+		// Name should match Kubernetes naming conventions - validated based on Kubernetes DNS1123 Subdomain rules.
+		if errStrings := validation.IsDNS1123Subdomain(runtimeExtension.Name); len(errStrings) > 0 {
+			errs = append(errs, errors.Errorf("Name %s is not valid: %s", runtimeExtension.Name, errStrings))
 		}
+
+		// Timeout should be a positive integer and should be under 600 seconds (10 minutes)
+		if *runtimeExtension.TimeoutSeconds < 0 && *runtimeExtension.TimeoutSeconds > 600 {
```
Comment (ykakarap) [inline exp/runtime/internal/controllers/extension_controller.go:220]: 10 minutes for maximum timeout seems too long. The admission webhooks in k8s allow a maximum timeout of 30s. 
How about using the same for our maximum as well?

File: exp/runtime/internal/controllers/extension_controller.go [hunk 3]
```diff
@@ -197,14 +200,59 @@ func (r Reconciler) discoverExtension(ctx context.Context, extension *runtimev1.
...
+		// TODO: Default Timeout and FailPolicy here.
+		if runtimeExtension.FailurePolicy == nil {
+			defaultFailPolicy := runtimev1.FailurePolicyFail
+			runtimeExtension.FailurePolicy = &defaultFailPolicy
+		}
+
+		// FIXME: How should we handle timeout is nil or zero?
+		// If TimeoutSeconds is not defined set to zero.
+		if runtimeExtension.TimeoutSeconds == nil {
+			runtimeExtension.TimeoutSeconds = pointer.Int32(0)
```
Comment (ykakarap) [inline exp/runtime/internal/controllers/extension_controller.go:252]: A 0s timeout means if the user does not set a timeout the network call will always timeout. The admission webhooks in k8s default to a timeout of 10s. We should also probably use the same timeout as default.

**Review by ykakarap — CHANGES_REQUESTED** [review]

**Comment by killianmuldoon** [comment]
Closing in favor of #https://github.com/kubernetes-sigs/cluster-api/pull/6578

**Pull request closed without merging**